### PR TITLE
[DO NOT MERGE] For SdW laptop

### PR DESCRIPTION
**Pull request by bardliao**
This is the difference between my branch and thesofproject/integration/soundwire-latest. I will rebase everyday and QA can take commits from here.

### CODE REVIEW

**Comment by plbossart** [comment]
@bardliao I think those parts are already mostly present on integration/soundwire-latest? It's only the apci-match and the link2 removal that aren't merged.

**Comment by bardliao** [comment]
> @bardliao I think those parts are already mostly present on integration/soundwire-latest? It's only the apci-match and the link2 removal that aren't merged.

@plbossart Yes, for today, but not yet yesterday :) Actually, rt715 DMIC3/4 support is not merged into soundwire-latest yet though it is already merged into realtek-sdw-codecs.

File: sound/soc/codecs/rt1308-sdw.c [hunk 1]
```diff
@@ -265,9 +265,6 @@ static int rt1308_update_status(struct sdw_slave *slave,
 	/* Update the status */
 	rt1308->status = status;
 
-	if (status == SDW_SLAVE_UNATTACHED)
-		rt1308->hw_init = false;
```
Comment (plbossart) [inline sound/soc/codecs/rt1308-sdw.c:292]: @shumingfan, I get that maybe for a 'SoftReset' due to a sync loss we don't need to reinitialize, but what does this have to do with resuming? On resume, you have to assume that the initializations are gone, we've cut the power to the codec.

Comment (shumingfan) [reply sound/soc/codecs/rt1308-sdw.c:292]: @plbossart I am not sure if HW engineer suggests the codec's power keeps or not. I also will check with the SW engineer who is in charge of SdW Windows driver next week.
I think we may use a register to check whether the codec's power disappeared ever.
To check this register, we decide whether to re-do the io_init function. How do you think?

Comment (plbossart) [reply sound/soc/codecs/rt1308-sdw.c:292]: @shumingfan, this case was added specifically for the suspend-resume case, where definitively we will lose state.
I just don't see how removing this if() case will improve things?

Comment (shumingfan) [reply sound/soc/codecs/rt1308-sdw.c:292]: @plbossart Rander reported an issue to me that the speaker has no sound output after the resume.
The speaker should output immediately after the resume because he does the headphone/speaker switch test. Due to the unattached state happened, the io_init process and the power-up of the playback process will be interleaved. 
That's why I suggest the io_init function doesn't do again when the unattached state happened.

**Comment by plbossart** [comment]
>   
> -	if (status == SDW_SLAVE_UNATTACHED)
> -		rt1308->hw_init = false;
> 
> @plbossart <https://github.com/plbossart> Rander reported an issue to me 
> that the speaker has no sound output after the resume.
> The speaker should output immediately after the resume because he does 
> the headphone/speaker switch test. Due to the unattached state happened, 
> the io_init process and the power-up of the playback process will be 
> interleaved.
> That's why I suggest the io_init function doesn't do again when the 
> unattached state happened.

Humm, I don't really understand why removing these lines would help?

Out of suspend, the codec has lost all its context so we do need to 
reinitialize it.

@RanderWang, please file a bug if you have a sequence that doesnt' work, 
so that we can figure things out.

**Comment by bardliao** [comment]
> - if (status == SDW_SLAVE_UNATTACHED) - rt1308->hw_init = false; @plbossart <https://github.com/plbossart> Rander reported an issue to me that the speaker has no sound output after the resume. The speaker should output immediately after the resume because he does the headphone/speaker switch test. Due to the unattached state happened, the io_init process and the power-up of the playback process will be interleaved. That's why I suggest the io_init function doesn't do again when the unattached state happened.
> Humm, I don't really understand why removing these lines would help? Out of suspend, the codec has lost all its context so we do need to reinitialize it. @RanderWang, please file a bug if you have a sequence that doesnt' work, so that we can figure things out.

@plbossart It is https://github.com/thesofproject/linux/issues/1373

**Comment by RanderWang** [comment]
> - if (status == SDW_SLAVE_UNATTACHED) - rt1308->hw_init = false; @plbossart <https://github.com/plbossart> Rander reported an issue to me that the speaker has no sound output after the resume. The speaker should output immediately after the resume because he does the headphone/speaker switch test. Due to the unattached state happened, the io_init process and the power-up of the playback process will be interleaved. That's why I suggest the io_init function doesn't do again when the unattached state happened.
> Humm, I don't really understand why removing these lines would help? Out of suspend, the codec has lost all its context so we do need to reinitialize it. @RanderWang, please file a bug if you have a sequence that doesnt' work, so that we can figure things out.

@plbossart   For this bug,  I found 1308 was reinitialized and the audio pipeline was set up completely and began to work but the initialization of 1308 was still not complete. The made no sound for 1308 stream.

If codec can keep the status, we don't need to reinitialize codec, or we may need to synchronize ASoC and codec initialization.

**Comment by plbossart** [comment]
@RanderWang we already have a synchronization with regmap, so if this doesn't happen we need to debug further. I can't accept a workaround if we don't know what the problem is.

**Comment by RanderWang** [comment]
> @RanderWang we already have a synchronization with regmap, so if this doesn't happen we need to debug further. I can't accept a workaround if we don't know what the problem is.

yes,  @shumingfan is working on it.

**Comment by RanderWang** [comment]
> @RanderWang we already have a synchronization with regmap, so if this doesn't happen we need to debug further. I can't accept a workaround if we don't know what the problem is.

for 1308,  we only do regmap synchronization in PM function now.
The error sequence is:
(1) aplay triggered master and 1308 resume
(2) master received unattach event, so set hw_init = false
(3) master received attach event, so began to init hw
(4) 1308 was initializing io (there is a msleep(100) and msleep(20) in rt1308_io_init)
(5) pcm_open, hw_params, trigger start (in another thread)
(6) initialization was complete.

My idea was we need to make step(5) to wait for (6)

**Comment by shumingfan** [comment]
@plbossart I am trying to make a patch that handles the unattached state happened.
I follow the method which is the same with Windows. 
The driver will read a certain register to check whether the system cuts the power to codec ever.  
If the power of codec disables in S3, the content of a certain register will lose. 
The codec driver will check this condition when the unattached state happened.
And it will check this condition in the resume function.
When the certain register loses the content, the driver will do the io_init function again.
The io_init function needs to finish before the regcache_sync in resume function.

But, I meet a problem and Bard also knows that as the comment below.
https://github.com/thesofproject/linux/issues/1344#issuecomment-547777233
I will look into this problem with @bardliao .

**Comment by plbossart** [comment]
@shumingfan the idea of checking if you need to reinit based on hw_status is fine with me. 

I provided my comments on the attach race condition, something isn't quite right in @bardliao 's explanations. The code could very well be wrong, i just don't see the issue clearly and the fix either.

@RanderWang I can see that we have a potential race between the codec finishing the inits, and the application starting playback. But is this an error?  This should not result in an IO issues, so what is the 'error' you are referring to?
If we do need to wait, then we should have a wait_for_completion in the DAI .open, where we'd wait for the init to be complete, eg.

1. rt711_resume: init_completion(&slave->initialization_complete)

2. rt711 dai open: wait_for_completion(&slave->initialization_complete). That is guaranteed by the framework to happen after the init_completion.

3.  hw_init: complete(&slave->initialization_complete)  (after everything is restored)

2 and 3 could happen in any order, that's fine.

**Comment by RanderWang** [comment]
> @RanderWang I can see that we have a potential race between the codec finishing the inits, and the application starting playback. But is this an error? This should not result in an IO issues, so what is the 'error' you are referring to?

@plbossart  sorry,  it should be when the error (no sound) occurs, the calling sequence is xxxx.
